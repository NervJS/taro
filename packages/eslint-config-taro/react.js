module.exports = Object.assign({}, require('./index'), {
  settings: {
    react: {
      pragma: 'React',
      version: 'detect'
    }
  }
})
