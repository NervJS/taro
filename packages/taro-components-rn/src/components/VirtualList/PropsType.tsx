import React, { ComponentType } from 'react'
import { StyleProp, ViewStyle, NativeSyntheticEvent, NativeScrollEvent } from 'react-native'
export interface ViewProps {
  style?: StyleProp<ViewStyle>;
  children?: React.ReactNode;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export interface VirtualListProps<T> {
  /** 列表的高度。 */
  height: string | number;
  /** 列表的宽度。 */
  width: string | number;
  /** 列表的长度 */
  itemCount: number;
  /** 渲染数据 */
  itemData: any[];
  /** 列表单项的大小，垂直滚动时为高度，水平滚动时为宽度。 */
  itemSize: number;
  /** 初始滚动偏移值，水平滚动影响 scrollLeft，垂直滚动影响 scrollTop。 */
  initialScrollOffset?: number;
  /** 列表内部容器组件类型，默认值为 View。 */
  innerElementType?: ComponentType;
  /** 滚动方向。vertical 为垂直滚动，horizontal 为平行滚动。默认为 vertical。 */
  layout?: 'vertical' | 'horizontal';
  /** 列表滚动时调用函数 */
  onScroll?: (event: VirtualListEvent<VirtualListProps.onScrollDetail>) => void;
  /** 调用平台原生的滚动监听函数。 */
  onScrollNative?: (event: NativeSyntheticEvent<NativeScrollEvent>) => void;
  /** 在可视区域之外渲染的列表单项数量，值设置得越高，快速滚动时出现白屏的概率就越小，相应地，每次滚动的性能会变得越差。 */
  overscanCount?: number;
  /** 是否注入 isScrolling 属性到 children 组件。这个参数一般用于实现滚动骨架屏（或其它 placeholder） 时比较有用。 */
  useIsScrolling?: boolean;
  children: ComponentType<{
    /** 单项的样式，样式必须传入组件的 style 中 */
    style?: Record<string, unknown>;
    /** 组件渲染的数据 */
    data: any;
    /** 单个列表组件数据 */
    item: any;
    /** 组件渲染数据的索引 */
    index: number;
    /** 组件是否正在滚动，当 useIsScrolling 值为 true 时返回布尔值，否则返回 undefined */
    // isScrolling?: boolean
  }>;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
interface VirtualListEvent<T> {
  /** 滚动方向，可能值为 forward 往前， backward 往后。 */
  scrollDirection: 'forward' | 'backward';
  /** 滚动距离 */
  scrollOffset: number;
  /** 当滚动是由 scrollTo() 或 scrollToItem() 调用时返回 true，否则返回 false */
  scrollUpdateWasRequested: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-namespace
declare namespace VirtualListProps {
  interface onScrollDetail {
    clientWidth: number;
    clientHeight: number;
  }
}
