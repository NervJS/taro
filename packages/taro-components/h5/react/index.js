import reactifyWc from './utils/reactify-wc'
import ReactInput from './components/input'

export const View = reactifyWc('taro-view-core')
export const Icon = reactifyWc('taro-icon-core')
export const Progress = reactifyWc('taro-progress-core')
export const RichText = reactifyWc('taro-rich-text-core')
export const Text = reactifyWc('taro-text-core')
export const Button = reactifyWc('taro-button-core')
export const Checkbox = reactifyWc('taro-checkbox-core')
export const CheckboxGroup = reactifyWc('taro-checkbox-group-core')
export const Editor = reactifyWc('taro-editor-core')
export const Form = reactifyWc('taro-form-core')
export const Input = ReactInput
export const Label = reactifyWc('taro-label-core')
export const Picker = reactifyWc('taro-picker-core')
export const PickerView = reactifyWc('taro-picker-view-core')
export const PickerViewColumn = reactifyWc('taro-picker-view-column-core')
export const Radio = reactifyWc('taro-radio-core')
export const RadioGroup = reactifyWc('taro-radio-group-core')
export const Slider = reactifyWc('taro-slider-core')
export const Switch = reactifyWc('taro-switch-core')
export const CoverImage = reactifyWc('taro-cover-image-core')
export const Textarea = reactifyWc('taro-textarea-core')
export const CoverView = reactifyWc('taro-cover-view-core')
export const MovableArea = reactifyWc('taro-movable-area-core')
export const MovableView = reactifyWc('taro-movable-view-core')
export const ScrollView = reactifyWc('taro-scroll-view-core')
export const Swiper = reactifyWc('taro-swiper-core')
export const SwiperItem = reactifyWc('taro-swiper-item-core')
export const FunctionalPageNavigator = reactifyWc('taro-functional-page-navigator-core')
export const Navigator = reactifyWc('taro-navigator-core')
export const Audio = reactifyWc('taro-audio-core')
export const Camera = reactifyWc('taro-camera-core')
export const Image = reactifyWc('taro-image-core')
export const LivePlayer = reactifyWc('taro-live-player-core')
export const Video = reactifyWc('taro-video-core')
export const Map = reactifyWc('taro-map-core')
export const Canvas = reactifyWc('taro-canvas-core')
export const Ad = reactifyWc('taro-ad-core')
export const OfficialAccount = reactifyWc('taro-official-account-core')
export const OpenData = reactifyWc('taro-open-data-core')
export const WebView = reactifyWc('taro-web-view-core')
export const NavigationBar = reactifyWc('taro-navigation-bar-core')
export const Block = reactifyWc('taro-block-core')
export const CustomWrapper = reactifyWc('taro-custom-wrapper-core')
