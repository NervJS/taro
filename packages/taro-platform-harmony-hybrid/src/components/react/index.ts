/* eslint-disable simple-import-sort/exports */
export { Ad } from '@tarojs/components/lib/react'
export { AdCustom } from '@tarojs/components/lib/react'
export { AnimationVideo } from '@tarojs/components/lib/react'
export { AnimationView } from '@tarojs/components/lib/react'
export { ArCamera } from '@tarojs/components/lib/react'
export { Audio } from '@tarojs/components/lib/react'
export { AwemeData } from '@tarojs/components/lib/react'
export { Block } from '@tarojs/components/lib/react'
export { Button } from '@tarojs/components/lib/react'
export { Camera } from '@tarojs/components/lib/react'
export { Canvas } from '@tarojs/components/lib/react'
export { ChannelLive } from '@tarojs/components/lib/react'
export { ChannelVideo } from '@tarojs/components/lib/react'
export { Checkbox, CheckboxGroup } from '@tarojs/components/lib/react'
export { CommentDetail, CommentList } from '@tarojs/components/lib/react'
export { ContactButton } from '@tarojs/components/lib/react'
export { CoverImage } from '@tarojs/components/lib/react'
export { CoverView } from '@tarojs/components/lib/react'
export { CustomWrapper } from '@tarojs/components/lib/react'
export { Editor } from '@tarojs/components/lib/react'
export { FollowSwan } from '@tarojs/components/lib/react'
export { Form } from '@tarojs/components/lib/react'
export { FunctionalPageNavigator } from '@tarojs/components/lib/react'
export { Icon } from '@tarojs/components/lib/react'
export { Image } from '@tarojs/components/lib/react'
export { InlinePaymentPanel } from '@tarojs/components/lib/react'
export { Input } from '@tarojs/components/lib/react'
export { KeyboardAccessory } from '@tarojs/components/lib/react'
export { Label } from '@tarojs/components/lib/react'
// export { Label } from './components/label/label'
export { Lifestyle } from '@tarojs/components/lib/react'
export { Like } from '@tarojs/components/lib/react'
export { LivePlayer } from '@tarojs/components/lib/react'
export { LivePusher } from '@tarojs/components/lib/react'
export { Login } from '@tarojs/components/lib/react'
export { Lottie } from '@tarojs/components/lib/react'
export { Map } from '@tarojs/components/lib/react'
export { MovableArea, MovableView } from '@tarojs/components/lib/react'
export { NavigationBar } from '@tarojs/components/lib/react'
export { Navigator } from '@tarojs/components/lib/react'
export { OfficialAccount } from '@tarojs/components/lib/react'
export { OpenData } from '@tarojs/components/lib/react'
export { PageMeta } from '@tarojs/components/lib/react'
export { Picker, PickerGroup } from '@tarojs/components/lib/react'
export { PickerView, PickerViewColumn } from '@tarojs/components/lib/react'
export { Progress } from '@tarojs/components/lib/react'
export { PullToRefresh } from '@tarojs/components/lib/react'
export { Radio, RadioGroup } from '@tarojs/components/lib/react'
export { RichText } from '@tarojs/components/lib/react'
export { RtcRoom, RtcRoomItem } from '@tarojs/components/lib/react'
export { ScrollView } from '@tarojs/components/lib/react'
export { ShareElement } from '@tarojs/components/lib/react'
export { Slider } from '@tarojs/components/lib/react'
export { NativeSlot, Slot } from '@tarojs/components/lib/react'
export { Swiper, SwiperItem } from '@tarojs/components/lib/react'
export { Switch } from '@tarojs/components/lib/react'
export { Tabbar, TabItem } from '@tarojs/components/lib/react'
export { Tabs } from '@tarojs/components/lib/react'
export { Text } from '@tarojs/components/lib/react'
export { Textarea } from '@tarojs/components/lib/react'
export { Video, VideoControl, VideoDanmu } from '@tarojs/components/lib/react'
export { View } from '@tarojs/components/lib/react'
export { VoipRoom } from '@tarojs/components/lib/react'
export { WebView } from '@tarojs/components/lib/react'
// export * from './components/block/block'
// export * from './components/grid/grid'
// export * from './components/root-portal/root-portal'
// export * from './components/sticky-header/sticky-header'
// export * from './components/sticky-section/sticky-section'
// export * from './components/editor/editor'
// export { default as ListView } from './components/list/list'
// export { MatchMedia } from './components/match-media/match-media'
// export { default as PageContainer } from './components/page-container/page-container'
export { StickyHeader, StickySection, RootPortal, GridView, ListView, MatchMedia, PageContainer } from '@tarojs/components/lib/react'
