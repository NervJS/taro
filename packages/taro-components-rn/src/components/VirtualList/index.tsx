
import * as React from 'react'
import { FlatList, NativeSyntheticEvent, NativeScrollEvent, ListRenderItemInfo } from 'react-native'
import { noop } from '../../utils'
import { VirtualListProps } from './PropsType'
import { ScrollViewProps } from '../ScrollView/PropsType'

class _VirtualList extends React.Component<VirtualListProps<any> & ScrollViewProps<any>, any> {
  static defaultProps = {
    upperThreshold: 50,
    lowerThreshold: 50,
    enableBackToTop: false,
    layout: 'vertical',
    overscanCount: 1
  }

  scrollViewStartOffsetY = 0 // 用于记录手指开始滑动时ScrollView组件的Y轴偏移量，通过这个变量可以判断滚动方向
  scrollUpdateWasRequested = false

  state = {}

  $ref = React.createRef<any>()

  /**
   * 滑动开始回调事件
   * 注意：当刚刚开始滑动时，event.nativeEvent.contentOffset.y仍然是上次滑动的值，没有变化
   * @param event
   * @private
   */
  _onScrollBeginDrag = (e: NativeSyntheticEvent<NativeScrollEvent>): void => {
    const { layout } = this.props
    // event.nativeEvent.contentOffset.y表示Y轴滚动的偏移量
    const offsetY = e.nativeEvent.contentOffset[layout === 'vertical' ? 'y' : 'x']
    // 记录ScrollView开始滚动的Y轴偏移量
    this.scrollViewStartOffsetY = offsetY
  }

  _onScroll = (e: NativeSyntheticEvent<NativeScrollEvent>): void => {
    const { onScroll = noop, layout, onScrollNative = noop } = this.props
    onScrollNative(e)
    const offsetY = e.nativeEvent.contentOffset[layout === 'vertical' ? 'y' : 'x']
    onScroll({
      onScroll: this.scrollViewStartOffsetY < offsetY ? 'forward' : 'backward',
      scrollOffset: Math.abs(this.scrollViewStartOffsetY - offsetY),
      scrollUpdateWasRequested: this.scrollUpdateWasRequested
    })
  }

  scrollTo = ({ offset }: { offset: number }): void => {
    const node = this.$ref.current
    if (node) {
      (node as FlatList<any>).scrollToOffset({ offset, animated: !!this.props.scrollWithAnimation })
    }
  }

  scrollToItem = ({ offset }: { offset: number }): void => {
    const node = this.$ref.current
    if (node) {
      (node as FlatList<any>).scrollToOffset({ offset, animated: !!this.props.scrollWithAnimation })
    }
  }

  render(): JSX.Element {
    const { itemData, itemSize, layout, overscanCount, children, ...restProps } = this.props
    const itemStyle = layout === 'vertical' ? { height: itemSize } : { width: itemSize }
    const itemRow = ({ item, index, separators }: ListRenderItemInfo<Record<string, unknown>>) =>
      React.createElement(children, {
        data: itemData,
        key: index,
        index,
        item,
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        separators,
        style: {
          ...itemStyle
        }
      })
    return (
      <FlatList
        {...restProps}
        data={itemData}
        windowSize={overscanCount}
        horizontal={layout === 'horizontal'}
        ref={this.$ref}
        onScroll={this._onScroll}
        onScrollBeginDrag={this._onScrollBeginDrag}
        renderItem={itemRow}
        keyExtractor={(_item: Record<string, unknown>, index: number) => index + ''}
      />
    )
  }
}

export default _VirtualList
