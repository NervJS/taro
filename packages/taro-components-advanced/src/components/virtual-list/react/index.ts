import { ScrollView, View } from '@tarojs/components'
import React from 'react'

import List from './list'
import outerWrapper from './wrapper'

import type { VirtualListProps } from '../'
import type { IProps } from '../preset'

const VirtualList = React.forwardRef(function VirtualList (props: VirtualListProps, ref) {
  const {
    direction = 'ltr',
    outerElementType = ScrollView,
    innerElementType = View,
    itemElementType = View,
    initialScrollOffset = 0,
    overscanCount = 1,
    ...rest
  } = props as IProps

  if ('children' in rest) {
    console.warn('Taro(VirtualList): children props have been deprecated. ' + 'Please use the item props instead.')
    rest.item = rest.children as IProps['item']
  }
  if (rest.item instanceof Array) {
    console.warn('Taro(VirtualList): item should not be an array')
    rest.item = rest.item[0]
  }
  return React.createElement(List, {
    ref,
    ...rest,
    outerElementType,
    itemElementType,
    innerElementType,
    direction,
    initialScrollOffset,
    overscanCount,
    outerWrapper,
  })
})

export default VirtualList
