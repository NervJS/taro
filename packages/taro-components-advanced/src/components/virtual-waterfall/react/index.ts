import { ScrollView, View } from '@tarojs/components'
import React from 'react'

import Waterfall from './waterfall'
import outerWrapper from './wrapper'

import type { VirtualWaterfallProps } from '../'
import type { IProps } from '../preset'

const VirtualWaterfall = React.forwardRef(function VirtualWaterfall (props: VirtualWaterfallProps, ref) {
  const {
    outerElementType = ScrollView,
    innerElementType = View,
    itemElementType = View,
    initialScrollOffset = 0,
    overscanDistance = 50,
    ...rest
  } = props as IProps

  if ('children' in rest) {
    console.warn('Taro(VirtualWaterfall): children props have been deprecated. ' + 'Please use the item props instead.')
    rest.item = rest.children as IProps['item']
  }
  if (rest.item instanceof Array) {
    console.warn('Taro(VirtualWaterfall): item should not be an array')
    rest.item = rest.item[0]
  }
  return React.createElement(Waterfall, {
    ref,
    ...rest,
    outerElementType,
    itemElementType,
    innerElementType,
    initialScrollOffset,
    overscanDistance,
    outerWrapper,
  })
})

export default VirtualWaterfall
