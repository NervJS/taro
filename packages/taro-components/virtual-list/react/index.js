import React from 'react'
import FixedSizeList from './FixedSizeList'
import { ScrollView, View } from '@tarojs/components'

function convertPxToInt (style) {
  if (typeof style === 'string') {
    const str = style.toLowerCase()
    if (/px$/.test(str)) {
      return Number(str.replace(/px$/, ''))
    }
  }
  return style
}

const OuterScrollView = React.forwardRef(
  (props, ref) => {
    const { style, onScroll, onScrollNative, layout, ...rest } = props
    const handleScroll = event => {
      onScroll({
        ...event,
        currentTarget: {
          ...event.detail,
          clientWidth: convertPxToInt(style.width),
          clientHeight: convertPxToInt(style.height)
        }
      })

      if (typeof onScrollNative === 'function') {
        onScrollNative(event)
      }
    }

    return React.createElement(ScrollView, {
      ref,
      style,
      scrollY: layout === 'vertical',
      scrollX: layout === 'horizontal',
      onScroll: handleScroll,
      ...rest
    })
  }
)

const VirtualList = React.forwardRef((props, ref) => {
  const {
    direction = 'ltr',
    innerElementType = View,
    itemElementType = View,
    initialScrollOffset = 0,
    overscanCount = 1,
    ...rest
  } = props

  return React.createElement(FixedSizeList, {
    ref,
    ...rest,
    itemElementType,
    innerElementType,
    outerElementType: OuterScrollView,
    direction,
    initialScrollOffset,
    overscanCount
  })
})

export default VirtualList
