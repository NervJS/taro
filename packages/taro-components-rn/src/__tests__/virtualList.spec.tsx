import * as React from 'react'
import { render } from '@testing-library/react-native'
import { View } from 'react-native'
import VirtualList from '../components/VirtualList'
const Row = React.memo(props => {
  const ctx: any = props
  const { index, style, data } = ctx
  return (
    <View style={style}>
      Row {index} : {data[index]}
    </View>
  )
})

describe('VirtualList', () => {
  it('VirtualList render', () => {
    // @ts-ignore
    const { toJSON } = render(<VirtualList>{Row}</VirtualList>)
    expect(toJSON()).toMatchSnapshot()
  })
})
