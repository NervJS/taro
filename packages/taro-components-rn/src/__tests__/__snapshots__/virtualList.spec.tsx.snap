// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualList VirtualList render 1`] = `
<RCTScrollView
  enableBackToTop={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  keyExtractor={[Function]}
  lowerThreshold={50}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  upperThreshold={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={1}
>
  <View />
</RCTScrollView>
`;
